lambda/trmc.ml

```diff
@@ -0,0 +1,859 @@
+    ['a Choice.t] has an applicative structure, which is useful to write
+    the actual code transformation in the {!choice} function.
+*)
+module Choice = struct
+  type 'a t =
+    | Return of 'a
+    (** [Return t] means that there are no TRMC opportunities in the subterm [t]:
+        no matter which context we are in,
+        we should evaluate [t] and "return" it. *)
+    | Set of 'a settable
+    (** [Set t] represents a piece of code that does contain
+        TRMC opportunities: if the context allows, we can write parts of
+        it in destination-passing-style to turn non-tail calls into tail
+        calls. See the type [settable] below. *)
+
+  and 'a settable = {
+    dps: 'a Dps.t;
+    direct: unit -> 'a;
+    benefits_from_dps: bool;
+    explicit_tailcall_request: bool;
+  }
+  (**
+     A [settable] record represents code that may be written in destination-passing style
+     if its usage context allows it. More precisely:
+
+     - If the surrounding context is already in destination-passing
+       style, it has a destination available, we should produce the
+       code in [dps] -- a function parametrized over the destination.
+       It passes the ~tail boolean to indicate whether the produced code
+       is used in tail position.
+
+     - If the surrounding context is in direct style (no destination
+       is available), we should produce the fallback code from
+       [direct].
+
+      (Note: [direct] is also a function (on [unit]) to ensure that any
+      effects performed during code production will only happen once we
+      do know that we want to produce the direct-style code.)
+
+     The [benefits_from_dps] boolean indicates whether
+     the use of destination-passing-style is beneficial to this function,
+     in the sense that the [dps] version has strictly more recursive calls
+     to TRMC versions than the [direct] version. When [benefits_from_dps]
+     is false, there is no difference in number of TRMC sub-calls --
+     see the {!choice_makeblock} case.
+
+     The [explicit_tailcall_request] boolean is true when the user
+     used a [@tailcall] annotation on the optimizable callsite.
+     When one of several calls could be optimized, we expect that
+     exactly one of them will be annotated by the user, or fail
+     because the situation is ambiguous.
+  *)
+
+  let map_settable f s = {
+    direct = (fun () -> f (s.direct ()));
+    dps = Dps.map f s.dps;
+    benefits_from_dps = s.benefits_from_dps;
+    explicit_tailcall_request = s.explicit_tailcall_request;
+  }
+
+  let map f = function
+    | Return t -> Return (f t)
+    | Set s -> Set (map_settable f s)
+
+  let pair ((c1, c2) : 'a t * 'b t) : ('a * 'b) t =
+    match c1, c2 with
+    | Return v1, Return v2 -> Return (v1, v2)
+    | Set s1, Return v2 ->
+        Set (map_settable (fun v1 -> (v1, v2)) s1)
+    | Return v1, Set s2 ->
+        Set (map_settable (fun v2 -> (v1, v2)) s2)
+    | Set s1, Set s2 ->
```

Comment:
This is incorrect when the `Set` is `Linear_static`, as we discovered in offline discussion with @gasche.  The issue is that with a `Linear_static` `Set`, the `Return` component will forget the delayed constructor applications.

The conceptual fix is to convert the `Return` into a `Set` then apply the `Set` version, but it should be easier to convert the `Set` into a `Dynamic` directly (because that is what will happen in the `Set` version anyways).